File: 3_longest_substring_without_repeating_characters/lengthOfLongestSubstring.go
func lengthOfLongestSubstring(s string) int {
    n := len(s)

    if n == 0 {
        return 0
    }

    longest := 0
    hash := make(map[rune]int)
    j := 0

    for i, c := range s {
        if val, ok := hash[c]; ok {
            j = max(j, val + 1)
        }

        hash[c] = i
        longest = max(longest, i - j + 1)
    }

    return longest
}

func max(a, b int) int {
    if a > b {
        return a
    }

    return b
}

File: 3_longest_substring_without_repeating_characters/lengthOfLongestSubstring.cpp
class Solution {
    public:
        int lengthOfLongestSubstring(string s) {
            int n = s.length();
    
            if (n == 0)
                return 0;
    
            int longest = 0;
            unordered_map<char, int> map;
    
            int j = 0;
    
            for (int i = 0; i < n; i++) {
                if (map.contains(s[i])) {
                    j = max(j, map[s[i]] + 1);
                }
                map[s[i]] = i;
                longest = max(longest, i - j + 1);
            }
    
            return longest;
        }
    };